### PR TITLE
Added support for plugin diagnostics

### PR DESCRIPTION
**Pull request by kdembler**
Added diagnostic mode to lib.
If plugin is run without config from framework present, it will start in diagnostic mode.
Plugin authors also have an option to add flags, so they can change behaviour of the plugin based on presence/value of CLI args.

TODO:

- [x] Add documentation to README
- [x] Add tests

### CODE REVIEW

File: snap_plugin/v1/metric.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def __init__(self, namespace=[], version=None, tags={}, config={},
             self._namespace = Namespace(self._pb.Namespace, *namespace)
         else:
             raise TypeError("The 'namespace', kwarg requires a list or tuple "
-                            "of :obj:`snap_plugin.v1.namespace_element.Namesp"
+                            "of :obj:`snap_plugin.v1.namespace_element.NamespaceElement"
                             "aceElement.  (given: `{}`)"
```
Comment (PatrykMatyjasek) [inline snap_plugin/v1/metric.py]: redundant comment

File: snap_plugin/v1/metric.py [hunk 2]
```diff
@@ -100,15 +100,8 @@ def __init__(self, namespace=[], version=None, tags={}, config={},
...
-        if isinstance(config, (list, tuple)):
-            self._config = ConfigMap(pb=self._pb.Config, *config)
-        elif isinstance(config, dict):
-            self._config = ConfigMap(pb=self._pb.Config, **config)
-        elif isinstance(config, ConfigMap):
-            self._config = ConfigMap(config.items(), pb=self._pb.Config)
-        else:
-            raise TypeError("The 'config' kwarg requires a list, tuple or"
-                            " dict.  (given: `{}`)".format(type(config)))
+        self._set_config(config)
```
Comment (PatrykMatyjasek) [inline snap_plugin/v1/metric.py:105]: called method raises error, we should handle it

File: snap_plugin/v1/plugin.py [hunk 3]
```diff
@@ -180,43 +267,99 @@ def stop_plugin(self):
...
-        self._monitor.join()
-        sys.exit()
+                cls=_EnumEncoder
+            )+"\n")
+            sys.stdout.flush()
+            self._monitor.start()
+            self._monitor.join()
+            sys.exit()
+        elif self._mode == PluginMode.diagnostics and self.meta.type == PluginType.collector:
+            self._print_diagnostic()
```
Comment (PatrykMatyjasek) [inline snap_plugin/v1/plugin.py:327]: missing else

**Review by PatrykMatyjasek — CHANGES_REQUESTED** [review]

File: snap_plugin/v1/plugin.py (hunk 3 above)
Comment (kdembler) [reply snap_plugin/v1/plugin.py:327]: Fixed in 2a3ced9

File: snap_plugin/v1/metric.py (hunk 2 above)
Comment (kdembler) [reply snap_plugin/v1/metric.py:105]: Error should be unhandled, as it should be passed to user's scope to let him know he called constructor with wrong args. Added docstring info about errors raised in 2a3ced9 for better readability.

**Review by PatrykMatyjasek — APPROVED** [review]

**Comment by PatrykMatyjasek** [comment]
Please squash commits
LGTM

**Comment by kdembler** [comment]
@PatrykMatyjasek Squashed :+1:

**Pull request merged**